src/sinks/aws_cloudwatch_logs/service.rs

```diff
@@ -0,0 +1,329 @@
+        let now = Utc::now();
+        // Acceptable range of Event timestamps.
+        let age_range = (now - Duration::days(14)).timestamp_millis()
+            ..(now + Duration::hours(2)).timestamp_millis();
+
+        let mut events = events
+            .into_iter()
+            .filter(|e| age_range.contains(&e.timestamp))
+            .collect::<Vec<_>>();
+
+        // Sort by timestamp
+        events.sort_by_key(|e| e.timestamp);
+
+        info!(message = "Sending events.", events = %events.len());
+
+        let mut event_batches = Vec::new();
+        if events.is_empty() {
+            // This should happen rarely.
+            event_batches.push(Vec::new());
+        } else {
+            // We will split events into 24h batches.
+            // Relies on log_events being sorted by timestamp in ascending order.
+            while let Some(oldest) = events.first() {
+                let limit = oldest.timestamp + Duration::days(1).num_milliseconds();
+
+                if events.last().expect("Events can't be empty").timestamp <= limit {
+                    // Fast path.
+                    // In most cases the difference between oldest and newest event
+                    // is less than 24h.
+                    event_batches.push(events);
+                    break;
+                }
+
+                // At this point we know that an event older than the limit exists.
+                //
+                // We will find none or one of the events with timestamp==limit.
+                // In the case of more events with limit, we can just split them
+                // at found event, and send those before at with this batch, and
+                // those after at with the next batch.
+                let at = events
+                    .binary_search_by_key(&limit, |e| e.timestamp)
+                    .unwrap_or_else(|at| at);
+
+                // Can't be empty
+                let remainder = events.split_off(at);
+                event_batches.push(events);
+                events = remainder;
+            }
+        }
+
+        event_batches
+    }
+}
+
+impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
+    type Response = ();
+    type Error = CloudwatchError;
+    type Future = request::CloudwatchFuture;
+
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        if let Some(rx) = &mut self.token_rx {
+            match ready!(rx.poll_unpin(cx)) {
+                Ok(token) => {
+                    self.token = token;
+                    self.token_rx = None;
+                }
+                Err(_) => {
+                    // This case only happens when the `tx` end gets dropped due to an error
+                    // in this case we just reset the token and try again.
+                    self.token = None;
+                    self.token_rx = None;
+                }
+            }
```

Comment:
I think you can collapse this down to something like:

```rust
self.token = ready!(rx.poll_unpin(cx)).ok();
self.token_rx = None;
```

Reply:
fixed in ad628ee0140d02cb601f09e1145031ea7bd15895